lib/PuppeteerSharp/ExecutionContext.cs

```diff
@@ -22,44 +24,86 @@ public ExecutionContext(Session client, ContextPayload contextPayload, Func<dyna
         public string FrameId { get; internal set; }
         public bool IsDefault { get; internal set; }
 
-        public async Task<dynamic> Evaluate(string pageFunction, params object[] args)
+        public async Task<dynamic> EvaluateAsync(string script, params object[] args)
         {
-            var handle = await EvaluateHandleAsync(pageFunction, args);
+            var handle = await EvaluateHandleAsync(script, args);
             dynamic result = await handle.JsonValue();
             await handle.Dispose();
             return result;
         }
 
-        internal async Task<JSHandle> EvaluateHandleAsync(Func<object> pageFunction, object[] args)
+        public async Task<T> EvaluateAsync<T>(string script, params object[] args)
         {
-            throw new NotImplementedException();
+            var result = await EvaluateAsync(script, args);
+            return ((JToken)result).ToObject<T>();
         }
 
-        internal async Task<JSHandle> EvaluateHandleAsync(string pageFunction, object[] args)
+        internal async Task<JSHandle> EvaluateHandleAsync(string script, object[] args)
         {
-            if (!string.IsNullOrEmpty(pageFunction))
+            if (string.IsNullOrEmpty(script))
             {
-                dynamic remoteObject;
+                return null;
+            }
+
+            if (IsFunction(script))
```

Comment:
maybe expose `EvaluateFunctionAsync` & `EvaluateExpressionAsync` instead of trying to implement `IsFunction`?

Reply:
Love it. By doing that we'll expose to APIs as Puppeteer does.

Reply:
I thought about this and I think we should leave only one entry point here. It has no benefit for the user having 2. She can send us a script and we know what to do there.

Reply:
my issue with this is that a function can be written in using different syntaxes:
```js
// 1.
function() { return 1 + 2; }
// 2.
() => 1 + 2;
// 3.
arg => 1 + 2;
// 4.
(arg1, arg2) => 1 + 2;
// not a function but contains =>
document.querySelectorAll('div').forEach(e=> console.log(e.innerText))
```

Reply:
Good catch!